### PR TITLE
Site Editor: Redesign table views

### PR DESCRIPTION
**Pull request by SaxonF**
This creates a new Page component along with FilterBar and Table components. I've kept the design _mostly_ the same for this PR so that we can focus a follow up PR on updating design to reflect some of the more recent concepts, including the pattern management grid.

image

## Why?
These components are to be used by pages like the existing templates manage and the [upcoming library](https://github.com/WordPress/gutenberg/issues/50028) (replacing template parts). It's the first step towards resolving [this issue.](https://github.com/WordPress/gutenberg/issues/50430).

## How?
- Introduces a generic `Page` component that accepts title, subTitle, actions and children
- A FilterBar component which creates a set of filter inputs and filters an array of data
- A Table component with an API that closely mimics [TanStack Table](https://tanstack.com/table/v8/docs/examples/react/basic) which we can implement in future if needed
- Replaces the existing List component with a main router like component that renders either Template or Library (template_part) pages. I've separated these two due to the upcoming work around Library.

## Testing Instructions
- Open site editor
- Open templates or template parts
- Click manage
- Management views should work as expected

### CODE REVIEW

File: packages/edit-site/src/components/page-main/index.js
Comment (noisysocks) [inline packages/edit-site/src/components/page-main/index.js]: Maybe rename `PageMainTemplates` to `PageTemplates` and `PageMainLibrary` to `PageTemplateParts` so that we're using consistent vocabulary throughout the site editor.

File: packages/edit-site/src/components/page-main-library/index.js
Comment (noisysocks) [inline packages/edit-site/src/components/page-main-library/index.js]: When do we ever want an entity record that is being deleted to be returned by `useEntityRecords`? I am wondering if we can/should move this logic to within `useEntityRecords` so as to avoid the awkward `useEffect` here.

Same comment applies to `PageMainTemplates`.

File: packages/edit-site/src/components/page-main-library/index.js
Comment (noisysocks) [inline packages/edit-site/src/components/page-main-library/index.js]: Could remove this and just pass `setFilteredRecords` to `onFilter` yeah?

Same comment applies to `PageMainTemplates`.

File: packages/edit-site/src/components/page/header.js
Comment (noisysocks) [inline packages/edit-site/src/components/page/header.js]: Is this supposed to be `as`? Why not rely on `level`?

File: packages/edit-site/src/components/table/index.js
Comment (noisysocks) [inline packages/edit-site/src/components/table/index.js]: Does `tr`'s `role` not default to `'row'`? Same question for all the other `role`s in this file.

File: packages/edit-site/src/components/page-main-templates/index.js
Comment (noisysocks) [inline packages/edit-site/src/components/page-main-templates/index.js]: These need to be localised and should use real ellipsis characters (…).

File: packages/edit-site/src/components/page-main-library/index.js
Comment (noisysocks) [inline packages/edit-site/src/components/page-main-library/index.js]: I love this API design.

File: packages/edit-site/src/components/page-main-library/index.js
Comment (noisysocks) [inline packages/edit-site/src/components/page-main-library/index.js]: These need to be localised and should use real ellipsis characters (…).

File: packages/edit-site/src/components/page/style.scss
Comment (noisysocks) [inline packages/edit-site/src/components/page/style.scss]: Unnecessary as it's overridden by the line below.

File: packages/edit-site/src/components/page/style.scss
Comment (noisysocks) [inline packages/edit-site/src/components/page/style.scss]: Should use `$white`.

File: packages/edit-site/src/components/page/style.scss
Comment (noisysocks) [inline packages/edit-site/src/components/page/style.scss]: Is this needed? Z indexes should be defined in the `_z-index.scss` file.

File: packages/edit-site/src/components/table/index.js
Comment (noisysocks) [inline packages/edit-site/src/components/table/index.js]: Maybe `undefined` instead of `'none'` so that there is simply no inline style when there is no `column.maxWidth`?

File: packages/edit-site/src/components/page-main-templates/index.js
Comment (noisysocks) [inline packages/edit-site/src/components/page-main-templates/index.js]: Remember to delete this before merge.

File: packages/edit-site/src/components/filter-bar/style.scss
Comment (noisysocks) [inline packages/edit-site/src/components/filter-bar/style.scss]: Remember to delete all the commented out code before merge.

File: packages/edit-site/src/components/filter-bar/style.scss
Comment (noisysocks) [inline packages/edit-site/src/components/filter-bar/style.scss]: Can/should this be a grid unit?

File: packages/edit-site/src/components/filter-bar/index.js
Comment (noisysocks) [inline packages/edit-site/src/components/filter-bar/index.js]: Does this component need to be so flexible given that we only have one filterable property (`title.raw`)? Are we over-engineering?

File: packages/edit-site/src/components/filter-bar/index.js
Comment (noisysocks) [inline packages/edit-site/src/components/filter-bar/index.js]: Could we eliminate the need for this by having the user of `FilterBar` provide a function instead of a string?

```js
<FilterBar
 	data={ templates }
	onFilter={ handleFilter }
	properties={ [
		{
 			// Instead of this:
			key: 'title.raw',
			// Could be this:			
			value: ( template ) => template?.title?.raw,
 			type: 'search',
 			placeholder: 'Search templates...',
 			label: 'Search templates...',
 		},
 	] }
 />
```

File: package-lock.json
Comment (noisysocks) [inline package-lock.json:1]: Something's gone wrong here, I wouldn't expect `package-lock.json` to appear in the diff. Probably incorrect npm version or something like that.

**Review by noisysocks — COMMENTED** [review]
Haven't ran this yet but the code is looking real good. I love the design of `Table` and a lot of this PR reads like an ad for `@wordpress/components` – beautiful 🌈.

Not so sure about `FilterBar`. Seems like a lot of complexity for one filterable attribute. Maybe you know something I don't though e.g that lots more are coming. In my mind though we should just have the pages render a `SearchControl`.

**Comment by SaxonF** [comment]
@noisysocks thanks so much for the thorough review, I'll look at these tomorrow.

> Not so sure about FilterBar. Seems like a lot of complexity for one filterable attribute.

Yeah I probably got a bit carried away here :D I might pull FilterBar out into its own branch for later and use SearchControl for now.

File: packages/edit-site/src/components/page/header.js
Comment (SaxonF) [reply packages/edit-site/src/components/page/header.js]: Indeed it is meant to be `as` so uses h1 but we want level 4 styling.

File: packages/edit-site/src/components/table/index.js
Comment (SaxonF) [reply packages/edit-site/src/components/table/index.js]: Was following existing table structure but that also felt strange. I couldn't find any reason to add them so have removed.

File: packages/edit-site/src/components/page-main-library/index.js
Comment (SaxonF) [reply packages/edit-site/src/components/page-main-library/index.js]: This was logic copied over from existing `list` component. I'm unsure why it was added. @kevin940726 any thoughts on whether this is needed or not?

**Comment by SaxonF** [comment]
@noisysocks should be ready for a re-review. I removed filtering completely to break down the PR a bit.

File: packages/edit-site/src/components/table/index.js
Comment (kevin940726) [inline packages/edit-site/src/components/table/index.js:14]: Probably not a good idea to use indices as keys?

File: packages/edit-site/src/components/page/header.js
Comment (kevin940726) [inline packages/edit-site/src/components/page/header.js:22]: Any reason why we are using level 4? It seems like the heading here might be the most important one so maybe it makes sense to be level 1? 🤔

File: packages/edit-site/src/components/page/header.js
Comment (kevin940726) [inline packages/edit-site/src/components/page/header.js:19]: Nit: We can use [`<hgroup>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/hgroup) to group the heading with its subheading here. It doesn't seem to have any accessibility impact though so it's just a fun TIL moment and don't need to change 😅 .

File: packages/edit-site/src/components/page-templates/index.js
Comment (kevin940726) [inline packages/edit-site/src/components/page-templates/index.js:101]: Do we want to render a loading spinner or something like that as a follow-up?

Also for an empty list?

**Review by kevin940726 — COMMENTED** [review]
Thanks for trying it out! I like the simplicity of the API and the implementation! I left some questions around accessibility and some observations. I'll test it locally next!

File: packages/edit-site/src/components/page-main-library/index.js
Comment (kevin940726) [reply packages/edit-site/src/components/page-main-library/index.js]: I think that was added by @annziel in https://github.com/WordPress/gutenberg/pull/48449. I haven't looked into the PR though.

**Comment by jameskoster** [comment]
Pushed a couple of tiny fixes.

Also aligned the contents of the last item in each row `right`, as the action buttons seemed oddly placed. If we don't think this will translate well to other applications feel free to revert.

File: packages/edit-site/src/components/table/style.scss
Comment (kevin940726) [inline packages/edit-site/src/components/table/style.scss]: Probably better to move this line into `.edit-site-page-content` to avoid the hidden focus ring:

image

**Review by kevin940726 — COMMENTED** [review]
Tested it locally and LGTM! 👍 💯

File: packages/edit-site/src/components/page/header.js
Comment (SaxonF) [reply packages/edit-site/src/components/page/header.js:22]: Setting level is just for styling. It's still an H1 for markup reasons.

File: packages/edit-site/src/components/table/style.scss
Comment (SaxonF) [reply packages/edit-site/src/components/table/style.scss]: Nice pickup, adjusted.

File: packages/edit-site/src/components/page/header.js
Comment (kevin940726) [reply packages/edit-site/src/components/page/header.js:22]: Hmm, yeah, you're right! I think the API is confusing though, I'd imagine that setting `level` changes the `aria-level` attribute but it doesn't. It's a totally different issue though so feel free to ignore me 😆 .

**Comment by SaxonF** [comment]
@kevin940726 would you mind taking one final review of this and approving so we can merge :)

**Review by noisysocks — APPROVED** [review]
Lovely 👍

**Comment by kevin940726** [comment]
Sorry for the late reply! My only concern left is https://github.com/WordPress/gutenberg/pull/50766#discussion_r1222469181 which might come back and bite us from the back 😅 . Otherwise this is good stuff! And I'm perfectly okay if we want to handle that issue separately too :).

**Pull request merged**
